shellplugin/wm.x11.impl/src/main/java/org/trinity/shellplugin/wm/x11/impl/protocol/icccm/AbstractCachedProtocol.java

```diff
@@ -76,50 +77,36 @@ public int getProtocolAtomId() {
 		return this.protocolAtomId;
 	}
 
-	public ListenableFuture<Void> addProtocolListener(	@Nonnull final DisplaySurface xWindow,
+	public CompletableFuture<Void> addProtocolListener(	@Nonnull final DisplaySurface xWindow,
 														@Nonnull final ProtocolListener<P> listener,
 														@Nonnull final ExecutorService executor) {
-		return this.displayExecutor.submit(new Callable<Void>() {
-			@Override
-			public Void call() throws Exception {
-				AsyncListenable listeners = AbstractCachedProtocol.this.listenersByWindow.get(xWindow);
-				if (listeners == null) {
-					listeners = new AsyncListenableEventBus(MoreExecutors.sameThreadExecutor());
-					listenersByWindow.put(	xWindow,
-											listeners);
-				}
-				listeners.register(	listener,
-									executor);
-				return null;
+		return CompletableFuture.supplyAsync(() -> {
+			AsyncListenable listeners = AbstractCachedProtocol.this.listenersByWindow.get(xWindow);
+			if (listeners == null) {
+				listeners = new AsyncListenableEventBus(MoreExecutors.sameThreadExecutor());
+				listenersByWindow.put(	xWindow,
+										listeners);
 			}
-		});
+			listeners.register(	listener,
+								executor);
+			return null;
+		}, displayExecutor);
 	}
 
-	public ListenableFuture<Optional<P>> get(@Nonnull final DisplaySurface xWindow) {
+	public CompletableFuture<Optional<P>> get(@Nonnull final DisplaySurface xWindow) {
 
-		final ListenableFuture<Optional<P>> protocolFuture = this.displayExecutor.submit(new Callable<Optional<P>>() {
-			@Override
-			public Optional<P> call() throws Exception {
-				return AbstractCachedProtocol.this.protocolCache.get(xWindow);
+		final CompletableFuture<Optional<P>> protocolFuture = CompletableFuture.supplyAsync(() -> {
+			return AbstractCachedProtocol.this.protocolCache.get(xWindow);
```

Comment:
Why did you say the task won't be executed by the display executor? The second argument of `CompletableFuture.supplyAsync` is the `displayExecutor`, so the provided task will be executed on the given `displayExecutor`.